File: leetcode/525-contiguous-array/prefix.py
# Let n be the length of nums
#
# Time: O(n)
# Space: O(n)

from collections import defaultdict


class Solution:
    def findMaxLength(self, nums: list[int]) -> int:
        prefix = [0]
        acc = 0
        for n in nums:
            acc += 1 if n == 0 else -1
            prefix.append(acc)

        locs = defaultdict(list)
        for i, s in enumerate(prefix):
            locs[s].append(i)

        maxi = 0
        for _, loc in locs.items():
            maxi = max(maxi, loc[-1] - loc[0])
        return maxi

File: leetcode/525-contiguous-array/prefix.go
package onlinejudge

func findMaxLength(nums []int) int {
	prefix := []int{0}
	acc := 0
	for _, n := range nums {
		if n == 0 {
			acc++
		} else {
			acc--
		}
		prefix = append(prefix, acc)
	}

	locs := map[int][]int{}
	for i, p := range prefix {
		locs[p] = append(locs[p], i)
	}

	max := 0
	for _, loc := range locs {
		current := loc[len(loc)-1] - loc[0]
		if current > max {
			max = current
		}
	}
	return max
}
